scala/src/main/org/apache/spark/streaming/api/kafka/ParallelKafkaDataFetcherDStream.scala

```diff
@@ -0,0 +1,270 @@
+      }
+    }
+    (rdd, totalMessageCount)
+  }
+
+  private def initialize(): Unit = {
+    if (!initialized) {
+      if (OffsetsFetcher == null) {
+        OffsetsFetcher = new KafkaOffsetsFetcher(
+          context, fetcherStateSnapshot, refreshOffsetsInterval, kafkaParams, topics, fromOffsets)
+        OffsetsFetcher.start()
+      }
+
+      if (precomputeCache == null) {
+        precomputeCache = new RddPreComputeCache[R](context.sparkContext,
+          s"ParallelKafkaDataFetcher-$id", maxUnackedNum, spitLimit, numThreads,
+          StorageLevel(useDisk = false, useMemory = true, deserialized = true, 3))
+        precomputeCache.start()
+      }
+
+      if (offsetsScheduler == null) {
+        offsetsScheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+          "offsetsScheduler-" + UUID.randomUUID().toString.substring(0, 8))
+        offsetsScheduler.execute(new Runnable {
+          override def run(): Unit = {
+            while (true) {
+              val fetchResult = OffsetsFetcher.take()
+              preCompute(fetchResult)
+            }
+          }
+        })
+      }
+
+      initialized = true
+    }
+  }
+
+  override def compute(validTime: Time): Option[RDD[R]] = {
+    val preComputeOutput = precomputeCache.spit()
+    val (rdd, totalMessageCount) = processRddPreComputeOutput(validTime, preComputeOutput)
+    val inputInfo = StreamInputInfo(id, totalMessageCount)
+    ssc.scheduler.inputInfoTracker.reportInfo(validTime, inputInfo)
+    Some(rdd)
+  }
+
+  // override this method to un-persist precomputed RDDs that are no longer needed
+  override private[streaming] def clearMetadata(time: Time): Unit = {
+    if (preComputeOutputMap.containsKey(time)) {
+      for (record <- preComputeOutputMap.get(time)) {
+        record.rdd.unpersist(false)
+        precomputeCache.ackSeqNum(record.rddSeqNum)
+      }
+    }
+    super.clearMetadata(time)
+  }
+
+  override def start(): Unit = {
+    initialize()
+  }
+
+  override def stop(): Unit = {
+    if (OffsetsFetcher != null) {
+      OffsetsFetcher.stop()
+      OffsetsFetcher = null
+    }
+
+    if (precomputeCache != null) {
+      precomputeCache.stop()
+      precomputeCache = null
+    }
+
+    if (offsetsScheduler != null) {
+      offsetsScheduler.shutdown()
+      offsetsScheduler = null
+    }
+  }
+
+  private[streaming]
+  class ParallelKafkaDataFetcherCheckpointData extends DStreamCheckpointData(this) {
+
+    def batchForTime: mutable.HashMap[Time, List[KafkaOffsetsFetchResult]] = {
+      data.asInstanceOf[mutable.HashMap[Time, List[KafkaOffsetsFetchResult]]]
+    }
+
+    override def update(time: Time) {
+      batchForTime.clear()
+      for (t <- preComputeOutputMap.keySet().toList) {
+        if (generatedRDDs.contains(t)) {
+          batchForTime +=
+            (t -> preComputeOutputMap.get(t).map(_.userInfo.asInstanceOf[KafkaOffsetsFetchResult]))
+        } else {
+          preComputeOutputMap.remove(t)
```

Comment:
Can you explain this condition - when it happens and why you are removing from the map?